File: backend/app/parsers/x.py
"""X（Twitter）解析器。

链路：推文链接 → 提取推文 ID → vxtwitter 优先 / fxtwitter 兜底（公共 API，无需登录）
→ 仍失败且配置了 x_cookie 时用 Playwright 带 cookie 打开推文页抓取（登录墙/私密推文）。

媒体：X 的视频 / 动图（GIF）都是 mp4（无平台水印），图片在 pbs.twimg.com；
直链无防盗链、支持 Range，可直接播放与下载，无需本地代理。
"""
import json
import re

import httpx

from .base import MediaFile, ParseResult, PlatformParser

UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36"
)
H = {"User-Agent": UA, "Accept": "application/json"}

VX_API = "https://api.vxtwitter.com/i/status/{tid}"
FX_API = "https://api.fxtwitter.com/status/{tid}"

_TWEET_ID_RE = re.compile(
    r"(?:x|twitter)\.com/(?:[^/?#]+/status/|i/status/)(\d{15,20})"
)
_RAW_ID_RE = re.compile(r"\b(\d{15,20})\b")


def _ensure_video_tag(url: str) -> str:
    """video.twimg.com 视频直链需带 tag 签名参数（如 tag=29），否则返回 403。"""
    if "tag=" in url:
        return url
    sep = "&" if "?" in url else "?"
    return f"{url}{sep}tag=29"


class XParser(PlatformParser):
    platform = "x"

    def parse(self, url: str, remove_platform_wm: bool = True) -> ParseResult:
        tid = self._extract_tweet_id(url)
        data = self._fetch_public(tid)
        if data is None:
            data = self._fetch_with_cookie(tid)
        if data is None:
            from ..config import load_config

            hint = ""
            if load_config().get("x_cookie"):
                hint = "（登录态可能已过期，请在设置中重新打开 X 登录）"
            raise RuntimeError(
                f"X 解析失败：该推文可能已删除、仅登录可见，或第三方解析服务暂时不可用{hint}"
            )
        return self._to_result(data, tid)

    # ---- ID 提取 ----

    @staticmethod
    def _extract_tweet_id(text: str) -> str:
        m = _TWEET_ID_RE.search(text)
        if m:
            return m.group(1)
        m = _RAW_ID_RE.search(text)
        if m:
            return m.group(1)
        raise RuntimeError("无法从链接中提取推文 ID，请确认是 X/Twitter 的推文链接")

    # ---- 公共解析（vxtwitter 优先，fxtwitter 兜底） ----

    @staticmethod
    def _fetch_public(tid: str) -> dict | None:
        try:
            r = httpx.get(VX_API.format(tid=tid), headers=H, follow_redirects=True, timeout=20)
            if r.status_code == 200:
                d = r.json()
                if d.get("media_extended") or d.get("mediaURLs"):
                    return {"source": "vxtwitter", "data": d}
        except Exception:
            pass
        try:
            r = httpx.get(FX_API.format(tid=tid), headers=H, follow_redirects=True, timeout=20)
            if r.status_code == 200:
                d = r.json()
                t = d.get("tweet") or {}
                if t.get("media") and t["media"].get("all"):
                    return {"source": "fxtwitter", "data": t}
        except Exception:
            pass
        return None

    # ---- 登录墙兜底（Playwright + x_cookie） ----

    def _fetch_with_cookie(self, tid: str) -> dict | None:
        from ..config import load_config

        cookie_str = (load_config().get("x_cookie") or "").strip()
        if not cookie_str:
            return None
        try:
            from playwright.sync_api import sync_playwright
        except ImportError:
            return None

        captured: dict = {}
        with sync_playwright() as p:
            browser = p.chromium.launch(headless=True)
            ctx = browser.new_context(user_agent=UA)
            try:
                cookies = json.loads(cookie_str)
                if isinstance(cookies, list):
                    ctx.add_cookies(cookies)
            except Exception:
                pass  # 非法 cookie 数据则直接走无 cookie 流程

            def on_response(resp):
                if "TweetResultByRestId" in resp.url:
                    try:
                        captured["graphql"] = resp.json()
                    except Exception:
                        pass

            page = ctx.new_page()
            page.on("response", on_response)
            try:
                page.goto(f"https://x.com/i/status/{tid}", wait_until="domcontentloaded", timeout=45000)
                for _ in range(20):
                    if captured.get("graphql"):
                        break
                    page.wait_for_timeout(2000)
            except Exception:
                pass
            ctx.close()
            browser.close()

        gql = captured.get("graphql")
        if not gql:
            return None
        return {"source": "graphql", "data": self._extract_from_graphql(gql)}

    @staticmethod
    def _extract_from_graphql(gql: dict) -> dict | None:
        """从 GraphQL TweetResultByRestId 响应中提取媒体列表（vxtwitter 兼容结构）。"""
        try:
            res = gql["data"]["tweetResult"]["result"]
        except Exception:
            return None
        media = []
        details = res.get("media_details") or []
        if not details:
            legacy = (res.get("legacy") or {}).get("extended_entities") or {}
            for m in legacy.get("media") or []:
                item = {"id_str": m.get("id_str", ""), "type": m.get("type", "photo")}
                if m.get("type") == "photo":
                    item["url"] = (m.get("media_url_https") or "").replace("http://", "https://")
                else:
                    item["url"] = m.get("video_info", {}).get("variants", [{}])[0].get("url", "")
                item["thumbnail_url"] = (m.get("media_url_https") or "").replace("http://", "https://")
                item["duration_millis"] = m.get("video_info", {}).get("duration_millis", 0)
                media.append(item)
        else:
            for m in details:
                item = {
                    "id_str": m.get("id", ""),
                    "type": m.get("type", "photo"),
                    "thumbnail_url": m.get("media_url_https") or m.get("thumbnail_url") or "",
                }
                if m.get("type") == "photo":
                    item["url"] = m.get("media_url_https", "")
                else:
                    variants = m.get("video", {}).get("variants") or []
                    url = ""
                    for v in sorted(variants, key=lambda v: v.get("bitrate") or 0, reverse=True):
                        if v.get("content_type") == "video/mp4":
                            url = v["url"]
                            break
                    item["url"] = url
                    item["duration_millis"] = m.get("video", {}).get("duration_millis", 0)
                media.append(item)
        if not media:
            return None
        text = (res.get("legacy") or {}).get("full_text") or (res.get("legacy") or {}).get("text") or ""
        return {"media_extended": media, "text": text}

    # ---- 转 ParseResult ----

    @staticmethod
    def _to_result(data: dict, tid: str) -> ParseResult:
        source = data["source"]
        if source == "vxtwitter":
            d = data["data"]
            items = d.get("media_extended") or []
            title = (d.get("text") or "").strip() or f"推文 {tid}"
        elif source == "fxtwitter":
            d = data["data"]
            items = (d.get("media") or {}).get("all") or []
            title = (d.get("text") or "").strip() or f"推文 {tid}"
        else:  # graphql（已转成 vxtwitter 兼容结构）
            d = data["data"]
            items = d.get("media_extended") or []
            title = (d.get("text") or "").strip() or f"推文 {tid}"

        files = []
        for it in items:
            mtype = (it.get("type") or "photo").lower()
            url = (it.get("url") or "").strip()
            if not url:
                continue
            if mtype in ("video", "gif"):
                # X 的动图（GIF）是无声 mp4，统一按视频处理；视频直链需带 tag 签名参数否则 403
                files.append(
                    MediaFile("video", _ensure_video_tag(url), cover=(it.get("thumbnail_url") or None))
                )
            else:  # photo / 图集
                files.append(MediaFile("image", url))
        if not files:
            raise RuntimeError("该推文没有可下载的视频/图片内容")
        return ParseResult("x", title, "video" if files[0].kind == "video" else "image", files)


# ---- 登录引导（应用内弹出浏览器登录 X，自动保存 cookie） ----

# 登录流程全局状态（单实例足够）：idle / running / done / error
_login_state: dict = {"status": "idle", "error": None}


def start_x_login() -> dict:
    """启动系统 Edge（调试端口）打开 X 登录页，后台轮询登录结果并保存 cookie。

    使用固定登录档案目录：登录态长期保留，下次打开登录窗口直接复用，
    无需反复输入账号密码。
    """
    import os
    import socket
    import subprocess
    import threading
    import time

    from ..config import APP_DIR

    if _login_state["status"] == "running":
        return {"ok": False, "error": "已有登录流程正在进行中"}
    _login_state.update(status="running", error=None)

    # 关闭上次残留的登录实例（同一固定档案），避免档案被占用导致启动失败
    subprocess.run(
        [
            "powershell", "-NoProfile", "-Command",
            "Get-CimInstance Win32_Process -Filter \"Name='msedge.exe'\" | "
            "Where-Object { $_.CommandLine -match 'x_login_profile' } | "
            "ForEach-Object { Stop-Process -Id $_.ProcessId -Force -ErrorAction SilentlyContinue }",
        ],
        capture_output=True,
        timeout=20,
    )
    time.sleep(1)

    # 找可用端口
    s = socket.socket()
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()

    profile_dir = APP_DIR / "x_login_profile"
    profile_dir.mkdir(parents=True, exist_ok=True)
    edge = None
    for cand in (
        r"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe",
        r"C:\Program Files\Microsoft\Edge\Application\msedge.exe",
    ):
        if os.path.exists(cand):
            edge = cand
            break
    if not edge:
        _login_state.update(status="error", error="未找到 Edge 浏览器，请改用设置中手动粘贴 Cookie")
        return {"ok": False, "error": "未找到 Edge 浏览器"}

    subprocess.Popen(
        [
            edge,
            f"--remote-debugging-port={port}",
            f"--user-data-dir={profile_dir}",
            "--no-first-run",
            "--no-default-browser-check",
            "--disable-blink-features=AutomationControlled",
            "--disable-features=AutomationControlled",
            "https://x.com/i/flow/login",
        ],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    # 等待调试端口就绪
    deadline = time.time() + 20
    while time.time() < deadline:
        try:
            import httpx as _h

            _h.get(f"http://127.0.0.1:{port}/json/version", timeout=2)
            break
        except Exception:
            time.sleep(0.5)

    def _poll():
        try:
            result = wait_x_login(port, profile_dir)
        except Exception as exc:
            result = {"ok": False, "error": f"登录流程异常：{exc}"}
        if result.get("ok"):
            _login_state.update(status="done", error=None)
        else:
            _login_state.update(status="error", error=result.get("error"))

    threading.Thread(target=_poll, daemon=True).start()
    return {"ok": True, "message": "请在 Edge 窗口中完成 X 登录（建议用手机 X App 扫码，比输入账号更稳定）"}


def x_login_status() -> dict:
    """查询登录流程状态。"""
    return {
        "ok": _login_state["status"] == "done",
        "status": _login_state["status"],
        "error": _login_state.get("error"),
    }


def wait_x_login(port: int, profile: str, timeout: int = 300) -> dict:
    """轮询登录结果：出现 auth_token 即视为登录成功，保存 cookie 到配置并关闭浏览器。"""
    import json as _json
    import time

    from ..config import load_config, save_config

    deadline = time.time() + timeout
    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        return {"ok": False, "error": "缺少 playwright 依赖"}

    browser = None
    ctx = None
    try:
        with sync_playwright() as p:
            try:
                browser = p.chromium.connect_over_cdp(f"http://127.0.0.1:{port}")
            except Exception:
                return {"ok": False, "error": "无法连接登录浏览器（可能已关闭），请重试"}
            ctx = browser.contexts[0] if browser.contexts else browser.new_context()
            while time.time() < deadline:
                if not browser.is_connected():
                    return {"ok": False, "error": "登录窗口已关闭，登录未完成"}
                try:
                    cookies = ctx.cookies()
                except Exception:
                    time.sleep(2)
                    continue
                if any(c.get("name") == "auth_token" and c.get("value") for c in cookies):
                    save_config(load_config() | {"x_cookie": _json.dumps(cookies)})
                    return {"ok": True}
                time.sleep(2)
            return {"ok": False, "error": "登录超时，请重试"}
    finally:
        try:
            if ctx:
                ctx.close()
        except Exception:
            pass
        try:
            if browser:
                browser.close()
        except Exception:
            pass

File: backend-rust/src/parsers/x.rs
//! X（Twitter）解析器（对齐 Python backend/app/parsers/x.py）。
//!
//! 链路：推文链接 → 提取推文 ID → vxtwitter 优先 / fxtwitter 兜底（公共 API，无需登录）
//! → 登录墙兜底（配置了 x_cookie 时用浏览器带 cookie 抓取，见 cdp 模块）。
//!
//! 媒体：X 的视频 / 动图（GIF）都是 mp4（无平台水印），图片在 pbs.twimg.com；
//! 直链无防盗链、支持 Range，可直接播放与下载，无需本地代理。

use std::process::Command;
use std::time::{Duration, Instant};

use futures_util::{Sink, SinkExt, Stream, StreamExt};
use regex::Regex;
use serde_json::{json, Value};
use tokio_tungstenite::connect_async;
use tokio_tungstenite::tungstenite::Message;

use crate::config::{data_dir, load_config};
use crate::parsers::cdp::CdpBrowser;
use crate::schemas::{MediaFile, ParseResult};
use crate::tasks::manager::UA;
use crate::x_login::{current_target_ws, debug_port_ready};

const VX_API: &str = "https://api.vxtwitter.com/i/status/{tid}";
const FX_API: &str = "https://api.fxtwitter.com/status/{tid}";

fn tweet_id_re() -> Regex {
    Regex::new(r"(?:x|twitter)\.com/(?:[^/?#]+/status/|i/status/)(\d{15,20})").unwrap()
}

fn raw_id_re() -> Regex {
    Regex::new(r"\b(\d{15,20})\b").unwrap()
}

/// video.twimg.com 视频直链需带 tag 签名参数（如 tag=29），否则返回 403。
fn ensure_video_tag(url: &str) -> String {
    if url.contains("tag=") {
        return url.to_string();
    }
    let sep = if url.contains('?') { "&" } else { "?" };
    format!("{url}{sep}tag=29")
}

fn extract_tweet_id(text: &str) -> Result<String, String> {
    if let Some(c) = tweet_id_re().captures(text) {
        return Ok(c.get(1).unwrap().as_str().to_string());
    }
    if let Some(c) = raw_id_re().captures(text) {
        return Ok(c.get(1).unwrap().as_str().to_string());
    }
    Err("无法从链接中提取推文 ID，请确认是 X/Twitter 的推文链接".to_string())
}

/// 解析入口。返回 ParseResult。
pub async fn parse(url: &str, _remove_platform_wm: bool) -> Result<ParseResult, String> {
    let tid = extract_tweet_id(url)?;
    let data = fetch_public(&tid).await;
    let data = match data {
        Some(d) => d,
        None => {
            // 登录墙兜底：浏览器 + x_cookie（M3 CDP 阶段接入）；无 cookie 时直接失败
            fetch_with_cookie(&tid).await?
        }
    };
    to_result(&data, &tid)
}

// ---- 公共解析（vxtwitter 优先，fxtwitter 兜底） ----

async fn fetch_public(tid: &str) -> Option<Value> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(20))
        .build()
        .ok()?;
    // vxtwitter 优先
    let url = VX_API.replace("{tid}", tid);
    if let Ok(resp) = client.get(&url).headers(headers()).send().await {
        if resp.status() == 200 {
            if let Ok(d) = resp.json::<Value>().await {
                if d.get("media_extended").is_some() || d.get("mediaURLs").is_some() {
                    return Some(json!({"source": "vxtwitter", "data": d}));
                }
            }
        }
    }
    // fxtwitter 兜底
    let url = FX_API.replace("{tid}", tid);
    if let Ok(resp) = client.get(&url).headers(headers()).send().await {
        if resp.status() == 200 {
            if let Ok(d) = resp.json::<Value>().await {
                let t = d.get("tweet").cloned().unwrap_or(Value::Null);
                if let Some(media) = t.get("media") {
                    if media.get("all").is_some() {
                        return Some(json!({"source": "fxtwitter", "data": t}));
                    }
                }
            }
        }
    }
    None
}

fn headers() -> reqwest::header::HeaderMap {
    use reqwest::header::{HeaderMap, HeaderValue, ACCEPT, USER_AGENT};
    let mut m = HeaderMap::new();
    m.insert(USER_AGENT, HeaderValue::from_static(UA));
    m.insert(ACCEPT, HeaderValue::from_static("application/json"));
    m
}

// ---- 登录墙兜底（浏览器 + x_cookie；登录态由 x_login.rs 引导流程保存） ----

async fn fetch_with_cookie(tid: &str) -> Result<Value, String> {
    let cookie_str = load_config().x_cookie.trim().to_string();
    if cookie_str.is_empty() {
        return Err("X 解析失败：该推文可能已删除、仅登录可见，或第三方解析服务暂时不可用".to_string());
    }
    // 用固定登录档案启动 headless Edge：登录态持久化，无需反复登录
    let port = CdpBrowser::find_free_port();
    let profile = data_dir().join("x_login_profile");
    let edge = CdpBrowser::edge_path().ok_or("未找到 Edge 浏览器")?;
    let mut child = Command::new(&edge)
        .arg(format!("--remote-debugging-port={port}"))
        .arg(format!("--user-data-dir={}", profile.display()))
        .arg("--headless=new")
        .arg("--no-first-run")
        .arg("--no-default-browser-check")
        .arg("--disable-gpu")
        .arg("--disable-blink-features=AutomationControlled")
        .arg("--disable-features=AutomationControlled")
        .arg(format!("--user-agent={UA}"))
        .arg("about:blank")
        .spawn()
        .map_err(|e| format!("启动浏览器失败：{e}"))?;

    let result = async {
        // 等待调试端口就绪（最多 20s）
        let deadline = Instant::now() + Duration::from_secs(20);
        loop {
            if debug_port_ready(port).await {
                break;
            }
            if Instant::now() >= deadline {
                return Err("浏览器启动超时".to_string());
            }
            tokio::time::sleep(Duration::from_millis(300)).await;
        }
        let ws_url = current_target_ws(port).await.ok_or("无法获取浏览器页面调试地址")?;
        let cookies: Value = serde_json::from_str(&cookie_str).unwrap_or(Value::Array(vec![]));
        let gql = fetch_tweet_graphql(&ws_url, &cookies, tid).await?;
        let data = extract_from_graphql(&gql).ok_or_else(|| {
            "该推文没有可下载的视频/图片内容（或登录态已失效，请重新打开 X 登录）".to_string()
        })?;
        Ok(json!({"source": "graphql", "data": data}))
    }
    .await;

    let _ = child.kill();
    let _ = child.wait();
    result
}

/// 打开推文页并监听 TweetResultByRestId GraphQL 响应（对齐 Python _fetch_with_cookie 的 on_response）。
async fn fetch_tweet_graphql(ws_url: &str, cookies: &Value, tid: &str) -> Result<Value, String> {
    use base64::Engine;
    use base64::engine::general_purpose::STANDARD;

    let (mut ws, _) = connect_async(ws_url)
        .await
        .map_err(|e| format!("连接调试页面失败：{e}"))?;
    let mut next_id = 1u64;

    // 1. 注入登录 cookie（双保险：档案意外丢失登录态时恢复）。
    //    CDP 只接受标准字段：过滤掉 partitionKey 等私有字段，避免整个请求被拒。
    send_cmd(
        &mut ws,
        next_id,
        "Network.setCookies",
        json!({"cookies": sanitize_cookies(cookies)}),
    )
    .await?;
    next_id += 1;
    // 2. 开启 Network 域以接收响应事件
    send_cmd(&mut ws, next_id, "Network.enable", json!({})).await?;
    next_id += 1;
    // 3. 打开推文页（GraphQL 请求由页面自行发出）
    send_cmd(
        &mut ws,
        next_id,
        "Page.navigate",
        json!({"url": format!("https://x.com/i/status/{tid}")}),
    )
    .await?;
    next_id += 1;

    // 4. 监听响应事件（最多 45s），命中 TweetResultByRestId 即取响应体
    let deadline = Instant::now() + Duration::from_secs(45);
    loop {
        let msg = ws.next().await.ok_or("调试连接中断")?;
        let msg = msg.map_err(|e| format!("读取响应失败：{e}"))?;
        if let Message::Text(t) = msg {
            if let Ok(v) = serde_json::from_str::<Value>(&t) {
                let method = v.get("method").and_then(|m| m.as_str()).unwrap_or("");
                if method == "Network.responseReceived" {
                    let url = v.get("params").and_then(|p| p.get("response")).and_then(|r| r.get("url")).and_then(|u| u.as_str()).unwrap_or("");
                    if url.contains("TweetResultByRestId") {
                        let request_id = v.get("params").and_then(|p| p.get("requestId")).and_then(|r| r.as_str()).unwrap_or("");
                        eprintln!("[x] 捕获 TweetResultByRestId 响应 url_len={}", url.len());
                        let resp = send_cmd(
                            &mut ws,
                            next_id,
                            "Network.getResponseBody",
                            json!({"requestId": request_id}),
                        )
                        .await?;
                        next_id += 1;
                        let base64 = resp.get("result").and_then(|r| r.get("base64Encoded")).and_then(|b| b.as_bool()).unwrap_or(false);
                        let body = resp.get("result").and_then(|r| r.get("body")).and_then(|b| b.as_str()).unwrap_or("");
                        let text = if base64 {
                            STANDARD.decode(body).map_err(|e| format!("响应体解码失败：{e}"))?
                        } else {
                            body.as_bytes().to_vec()
                        };
                        let gql: Value = serde_json::from_slice(&text).map_err(|e| format!("GraphQL 响应解析失败：{e}"))?;
                        return Ok(gql);
                    }
                }
            }
        }
        if Instant::now() >= deadline {
            return Err("推文页面加载超时（登录态可能已失效，请重新打开 X 登录）".to_string());
        }
    }
}

/// 过滤 CDP cookie 为只含标准字段（name/value/domain/path/expires/httpOnly/secure/sameSite）。
fn sanitize_cookies(cookies: &Value) -> Value {
    let arr = cookies.as_array().cloned().unwrap_or_default();
    let cleaned: Vec<Value> = arr
        .into_iter()
        .map(|c| {
            let mut o = serde_json::Map::new();
            for k in ["name", "value", "domain", "path", "expires", "httpOnly", "secure", "sameSite"] {
                if let Some(v) = c.get(k) {
                    o.insert(k.to_string(), v.clone());
                }
            }
            Value::Object(o)
        })
        .collect();
    Value::Array(cleaned)
}

/// 发送 CDP 命令并等待同 id 响应（同一连接内保持会话，脚本注册跨命令有效）。
async fn send_cmd<S>(ws: &mut S, id: u64, method: &str, params: Value) -> Result<Value, String>
where
    S: Sink<Message, Error = tokio_tungstenite::tungstenite::Error>
        + Stream<Item = Result<Message, tokio_tungstenite::tungstenite::Error>>
        + Unpin,
{
    let cmd = json!({"id": id, "method": method, "params": params});
    ws.send(Message::Text(cmd.to_string().into()))
        .await
        .map_err(|e| format!("发送命令失败：{e}"))?;
    loop {
        let msg = ws.next().await.ok_or("调试连接中断")?;
        let msg = msg.map_err(|e| format!("读取响应失败：{e}"))?;
        if let Message::Text(t) = msg {
            if let Ok(v) = serde_json::from_str::<Value>(&t) {
                if v.get("id").and_then(|x| x.as_u64()) == Some(id) {
                    return Ok(v);
                }
            }
        }
    }
}

/// 从 GraphQL TweetResultByRestId 响应提取媒体（转成 vxtwitter 兼容结构）。
fn extract_from_graphql(gql: &Value) -> Option<Value> {
    let res = gql.get("data")?.get("tweetResult")?.get("result")?;
    let mut media: Vec<Value> = Vec::new();
    let details = res
        .get("media_details")
        .and_then(|m| m.as_array())
        .cloned()
        .unwrap_or_default();
    if details.is_empty() {
        // 旧结构：legacy.extended_entities.media
        let legacy = res.get("legacy").and_then(|l| l.get("extended_entities"));
        if let Some(arr) = legacy.and_then(|l| l.get("media")).and_then(|m| m.as_array()) {
            for m in arr {
                let mtype = m.get("type").and_then(|v| v.as_str()).unwrap_or("photo");
                let mut item = json!({
                    "id_str": m.get("id_str").and_then(|v| v.as_str()).unwrap_or(""),
                    "type": mtype,
                });
                let thumb = m
                    .get("media_url_https")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .replace("http://", "https://");
                if mtype == "photo" {
                    item["url"] = json!(thumb);
                } else {
                    let mut url = String::new();
                    if let Some(variants) = m
                        .get("video_info")
                        .and_then(|v| v.get("variants"))
                        .and_then(|v| v.as_array())
                    {
                        for v in variants {
                            if v.get("content_type").and_then(|c| c.as_str()) == Some("video/mp4") {
                                url = v.get("url").and_then(|u| u.as_str()).unwrap_or("").to_string();
                                break;
                            }
                        }
                    }
                    item["url"] = json!(url);
                }
                item["thumbnail_url"] = json!(thumb);
                media.push(item);
            }
        }
    } else {
        // 新结构：media_details
        for m in details {
            let mtype = m.get("type").and_then(|v| v.as_str()).unwrap_or("photo");
            let mut item = json!({
                "id_str": m.get("id").and_then(|v| v.as_str()).unwrap_or(""),
                "type": mtype,
            });
            let thumb = m
                .get("media_url_https")
                .and_then(|v| v.as_str())
                .or_else(|| m.get("thumbnail_url").and_then(|v| v.as_str()))
                .unwrap_or("")
                .to_string();
            if mtype == "photo" {
                item["url"] = json!(m.get("media_url_https").and_then(|v| v.as_str()).unwrap_or(""));
            } else {
                let mut url = String::new();
                if let Some(variants) = m
                    .get("video")
                    .and_then(|v| v.get("variants"))
                    .and_then(|v| v.as_array())
                {
                    for v in variants {
                        if v.get("content_type").and_then(|c| c.as_str()) == Some("video/mp4") {
                            url = v.get("url").and_then(|u| u.as_str()).unwrap_or("").to_string();
                            break;
                        }
                    }
                }
                item["url"] = json!(url);
            }
            item["thumbnail_url"] = json!(thumb);
            media.push(item);
        }
    }
    if media.is_empty() {
        return None;
    }
    let text = res
        .get("legacy")
        .and_then(|l| l.get("full_text"))
        .and_then(|v| v.as_str())
        .or_else(|| res.get("legacy").and_then(|l| l.get("text")).and_then(|v| v.as_str()))
        .unwrap_or("")
        .to_string();
    Some(json!({"media_extended": media, "text": text}))
}

// ---- 转 ParseResult ----

fn to_result(data: &Value, tid: &str) -> Result<ParseResult, String> {
    let source = data.get("source").and_then(|v| v.as_str()).unwrap_or("");
    let d = data.get("data").unwrap_or(&Value::Null);
    let (items, title) = if source == "vxtwitter" {
        let items = d.get("media_extended").cloned().unwrap_or(Value::Array(vec![]));
        let t = (d.get("text").and_then(|v| v.as_str()).unwrap_or("")).trim();
        let title = if t.is_empty() { format!("推文 {tid}") } else { t.to_string() };
        (items, title)
    } else if source == "fxtwitter" {
        let items = d
            .get("media")
            .and_then(|m| m.get("all"))
            .cloned()
            .unwrap_or(Value::Array(vec![]));
        let t = (d.get("text").and_then(|v| v.as_str()).unwrap_or("")).trim();
        let title = if t.is_empty() { format!("推文 {tid}") } else { t.to_string() };
        (items, title)
    } else if source == "graphql" {
        // graphql 已转成 vxtwitter 兼容结构（media_extended + text）
        let items = d.get("media_extended").cloned().unwrap_or(Value::Array(vec![]));
        let t = (d.get("text").and_then(|v| v.as_str()).unwrap_or("")).trim();
        let title = if t.is_empty() { format!("推文 {tid}") } else { t.to_string() };
        (items, title)
    } else {
        return Err("X 解析失败：第三方解析服务返回异常".to_string());
    };

    let mut files: Vec<MediaFile> = Vec::new();
    if let Some(arr) = items.as_array() {
        for it in arr {
            let mtype = it.get("type").and_then(|v| v.as_str()).unwrap_or("photo").to_lowercase();
            let url = it.get("url").and_then(|v| v.as_str()).unwrap_or("").trim().to_string();
            if url.is_empty() {
                continue;
            }
            if mtype == "video" || mtype == "gif" {
                // X 的动图（GIF）是无声 mp4，统一按视频处理；视频直链需带 tag 签名参数否则 403
                let cover = it.get("thumbnail_url").and_then(|v| v.as_str()).map(String::from);
                files.push(MediaFile {
                    kind: "video".to_string(),
                    url: ensure_video_tag(&url),
                    label: None,
                    cover,
                    image_url: None,
                });
            } else {
                // photo / 图集
                files.push(MediaFile {
                    kind: "image".to_string(),
                    url,
                    label: None,
                    cover: None,
                    image_url: None,
                });
            }
        }
    }
    if files.is_empty() {
        return Err("该推文没有可下载的视频/图片内容".to_string());
    }
    let media_type = if files[0].kind == "video" { "video" } else { "image" };
    Ok(ParseResult {
        platform: "x".to_string(),
        title,
        media_type: media_type.to_string(),
        files,
    })
}
